### ISSUE TITLE
Calls to functions in modules dynamically linked with loadDynamicLibrary much slower than with dlopen?

### ISSUE THREAD

**Issue by fibbo** [emscripten-core/emscripten#8302]
I have two test cases:
1. Load side modules asynchronously on demand directly in my C++ code with loadDynamicLibrary like this

```
void loadLibs() {
    EM_ASM({
        loadDynamicLibrary('output/side_module.wasm', {loadAsync: true, global: true, nodelete: true}).then(
            () => {
                Module.modulePromiseResolve() 
            });
        loadDynamicLibrary('output/side_module2.wasm', {loadAsync: true, global: true, nodelete: true}).then(
            () => {
                Module.module2PromiseResolve()
            });
    });
}
```
2. Use dlopen/dlsym

Running some test code turns out that calls to functions with dlopen are much faster than the ones from loadDynamicLibrary.

Particularly interesting to me is that this is even true for functions that are from the same module. Following function is in the main module and is called 6'000'000 times:

```
void NewDeleteMainModule()
{
    EM_ASM_TIME('NewDeleteMainModule'); // macro that calls console.time('timerName')
    for(size_t i = 0; i < SAMPLE; i++)
    {
        DataStruct* ds = new DataStruct;
        delete ds;
    }
    EM_ASM_TIMEEND('NewDeleteMainModule'); // macro that calls console.timeEnd('timerName')
}
```

Calling this in case `1.` takes roughly 770ms while in case `2.` it only takes 150ms. Is this to expected? Especially when calling functions from the same module?

**Comment by sbc100** [comment]
This is very interesting. Could be something do with the fact that SIDE_MODULE/MAIN_MODULE code is compiled with RELOCATABLE=1 I imagine effects codegen.  Certainly worth investigating.

It should be noted that compiling code with `-fPIC` in the native world (similar to what RELOCABLE=1 does) can also slow down the code somewhat as it introduces a level of indirection when calling global functions and accessing global data.

**Comment by fibbo** [comment]
I ran the tests again this morning in a new tab in the same browser (firefox developer ed.) but I was unable to reproduce the effect.

I deleted the cache and reloaded the old tab (had it still open and this one still took around 770ms even after deleting the cache (is this not enough?)) but in the new tab it was down to 150ms. So if there is a problem it seems a transient one and I cannot reproduce it any more. I will close this issue but I will keep an eye on it.
